### ISSUE TITLE
No such widget `.manydots-magic.orig.backward-delete-char'

### ISSUE THREAD

**Issue by Jomik** [knu/zsh-manydots-magic#2]
I am getting this issue after using zplug to load your code whenever I hit backspace.
No such widget `.manydots-magic.orig.backward-delete-char'
My config can be found here https://github.com/Jomik/dotfiles/blob/c92f8f78dabcb47bd275c90345907be3e26131db/home/zshrc

EDIT: Refer to actual commit and not master, for later usage if needed.

**Comment by mavam** [comment]
Looks like you're loading the plugin via zplug, Adding `defer:2` fixed it for me:

```
zplug 'knu/zsh-manydots-magic', use:manydots-magic, defer:2
```

**Comment by knu** [comment]
I'm not using zplug.  Is there anything I can do in manydots-magic?

**Comment by mavam** [comment]
Good question. The error message that crept up was probably occurring due to the order of how packages are loaded. The `defer` directive changes that order. In the README, you write:

> If you are enabling `url-quote-magic`, make sure to load `url-quote-magic` first and then load `manydots-magic`.

I assume that this was the culprit.

**Comment by Jomik** [comment]
@mavam I wasn't enabling url-quote-magic at the time.
I've however gone to using fish shell now.

**Comment by Virako** [comment]
I fixed the error with the next: https://unix.stackexchange.com/questions/368231/going-over-start-of-insert-action-in-z-shell-vi-mode

**Comment by twang817** [comment]
I've solved this problem for myself by loading it after zsh-autosuggestions:

```
zinit wait'0c' for \
    atload'_zsh_autosuggest_start' zsh-users/zsh-autosuggestions \
    pick'manydots-magic' compile'manydots-magic' knu/zsh-manydots-magic \
    ;
```

EDIT: I actually think this may be in conflict with fast-syntax-highlighting, not zsh-autosuggestions.

**Comment by protist** [comment]
It looks like the recent zsh update to 5.8.1 has broken this for me too. Every time I press backspace, I get the same error and cannot delete text. I tried reverting to zsh 5.8 and backspace works perfectly again.

I'm not using `zplug` or `url-quote-magic`. I don't really understand either fix proposed by the two previous commentors, but this might be something new and unrelated anyway.

**Comment by knu** [comment]
You need to respawn the shell after upgrading zsh as long as you use autoload.  It's not specific to this widget.

**Comment by protist** [comment]
@knu Sorry I'm not entirely sure what you mean. Do you mean if I open a new zsh shell, then it should work fine? If so, this doesn't work for me. I _am_ using the autoload method.

**Comment by protist** [comment]
I'm also using [zim](https://github.com/zimfw/zimfw), and I thought I should ~~update~~ reinstall this too. After ~~updating~~ reinstalling, I can use the backspace key fine now (I don't get the error), but backspace just does its "normal" role, and doesn't delete groups of `/..`

**Comment by knu** [comment]
@protist Do you happen to have zsh-autosuggestions installed?  I confirmed that zsh-autosuggestions overwrites LASTWIDGET that manydots-magit uses to know if backspace should delete the last `/..` part.  With it disabled, manydots-magic works fine for me including the backspace behavior.

**Comment by knu** [comment]
I think I'll loosen the conditions so that it only checks if LBUFFER ends with `../..` to see if backspace should delete the last `/..` part.

**Comment by protist** [comment]
@knu Yes, that is exactly it! If I disable zsh-autosuggestions, manydots-magic's backspace functionally does indeed work. Thanks for the detective work!

**Comment by twang817** [comment]
You guys might take a look at this hack that I did to solve the problem for me:

https://github.com/twang817/zsh-manydots-magic/commit/6525d31436fd3ea9ca0de877c7dccc5763f41e46

**Comment by protist** [comment]
Thank @twang817. I can confirm that with this fork I can both use zsh-autosuggestions and manydots-magic's backspace still works.

**Comment by kergoth** [comment]
> You guys might take a look at this hack that I did to solve the problem for me:
> 
> [twang817@6525d31](https://github.com/twang817/zsh-manydots-magic/commit/6525d31436fd3ea9ca0de877c7dccc5763f41e46)

Just as an FYI, with your fork I get this instead:
```
_zsh_autosuggest_bound_1_manydots-magic.self-insert:74: maximum nested function level reached; increase FUNCNEST?
```
